lib/active_merchant/billing/gateways/elavon.rb

```diff
@@ -352,30 +358,45 @@ def add_line_items(xml, level_3_data)
         }
       end
 
-      def add_stored_credential(xml, options)
+      def add_stored_credential(xml, payment_method, options)
+        return unless options[:stored_credential]
+
         network_transaction_id = options.dig(:stored_credential, :network_transaction_id)
-        case
-        when network_transaction_id.nil?
-          return
-        when network_transaction_id.to_s.include?('|')
+        if network_transaction_id.to_s.include?('|')
           oar_data, ps2000_data = options[:stored_credential][:network_transaction_id].split('|')
-          xml.ssl_oar_data oar_data unless oar_data.nil? || oar_data.empty?
-          xml.ssl_ps2000_data ps2000_data unless ps2000_data.nil? || ps2000_data.empty?
-        when network_transaction_id.to_s.length > 22
-          xml.ssl_oar_data options.dig(:stored_credential, :network_transaction_id)
-        else
-          xml.ssl_ps2000_data options.dig(:stored_credential, :network_transaction_id)
+          xml.ssl_oar_data oar_data unless oar_data.blank? || payment_method.is_a?(String)
+          xml.ssl_ps2000_data ps2000_data unless ps2000_data.blank? || payment_method.is_a?(String)
+        elsif network_transaction_id.to_s.length > 22
+          xml.ssl_oar_data network_transaction_id
+        elsif network_transaction_id.present?
+          xml.ssl_ps2000_data network_transaction_id
         end
+        xml.ssl_recurring_flag recurring_flag(options) if recurring_flag(options)
+        xml.ssl_approval_code options[:approval_code] if options[:approval_code] && !payment_method.is_a?(String) && !options[:ssl_token]
```

Comment:
Could we make this into a variable so it can also be used above? Where we check for `payment_method.is_a?(String)` in `ssl_oar_data` and `ssl_ps2000_data`. Maybe the the new variable could be called `used_payment_token`

Reply:
Good suggestion, I've done some refactoring of this method to clean it up in general which should help.